controller/src/main/java/org/jboss/as/controller/security/CredentialStoreClient.java

```diff
@@ -0,0 +1,190 @@
+     */
+    public CredentialStoreClient(final CredentialStore credentialStore, final String name, final String alias) {
+        this(credentialStore, name, alias, (Class) null);
+    }
+
+    /**
+     * Constructor to create instance of {@link CredentialStoreClient}.
+     * @param credentialStore actual {@link CredentialStore} used by this client
+     * @param name of credential store
+     * @param alias of credential in the store
+     * @param className of credential denoted by this alias
+     * @throws ClassNotFoundException when credential reference holding credential type which cannot be resolved using current providers
+     */
+    public CredentialStoreClient(final CredentialStore credentialStore, final String name, final String alias, final String className) throws ClassNotFoundException {
+        this(credentialStore, name, alias, toClass(className, credentialStore.getClass().getClassLoader()));
+    }
+
+    private static Class<? extends Credential> toClass(final String className, final ClassLoader classLoader) throws ClassNotFoundException {
+        return (Class<Credential>) Class.forName(className, true, classLoader);
+    }
+
+    /**
+     * Get the secret in form of clear text.
+     * @return secret as clear text
+     */
+    public char[] getSecret() {
+        if (isDestroyed()) {
+            return null;
+        }
+        PasswordCredential passwordCredential = (PasswordCredential) getCredential();
+        if (passwordCredential != null) {
+            if (passwordCredential.getPassword() instanceof ClearPassword) {
+                return ((ClearPassword) passwordCredential.getPassword()).getPassword();
+            } else {
+                ROOT_LOGGER.log(Logger.Level.DEBUG, ROOT_LOGGER.unsupportedCredentialType(name,
+                        passwordCredential.getPassword().getClass().getName()));
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get the {@link Credential} instance from the credential store.
+     * @return Credential instance of proper credential type specified if not available {@code null}
+     */
+    public Credential getCredential() {
+        if (isDestroyed() || alias == null) {
+            return null;
+        }
+        try {
+            if (type != null) {
+                return credentialStore.retrieve(alias, type);
+            } else {
+                return credentialStore.retrieve(alias, PasswordCredential.class);
+            }
+        } catch (UnsupportedCredentialTypeException | CredentialStoreException e) {
+            ROOT_LOGGER.log(Logger.Level.INFO, e);
+        }
+        return null;
+    }
+
+    /**
+     * Retrieve associated {@link CredentialStore}
+     * This method should perform check whether the {@link CredentialStore} can be returned to caller.
+     *
+     * @return associated {@link CredentialStore}
+     */
+    public CredentialStore getCredentialStore() {
+        // TODO: check caller if possible
```

Comment:
Is there a JIRA to take care of this?

Also, is it necessary to have this method public? Perhaps other code outside this PR uses it but the only use I see here is in this package to create a CredentialStoreClient from another CredentialStoreClient. So that can be done without public visibility.